ipld/car/src/util.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::error::Error;
+use cid::Cid;
+use std::io::Read;
+
+pub(crate) fn ld_read<R: Read>(mut buf_reader: &mut R) -> Result<(u64, Vec<u8>), Error> {
+    let l =
+        unsigned_varint::io::read_u64(&mut buf_reader).map_err(|e| Error::Other(e.to_string()))?;
+    let mut buf = Vec::with_capacity(l as usize);
+    buf_reader
+        .take(l)
+        .read_to_end(&mut buf)
+        .map_err(|e| Error::Other(e.to_string()))?;
+    Ok((l, buf))
+}
+
+pub(crate) fn read_node<R: Read>(buf_reader: &mut R) -> Result<(Cid, Vec<u8>), Error> {
+    let (_len, buf) = ld_read(buf_reader)?;
+    let (c, n) = read_cid(&buf)?;
+    Ok((c, buf[(n as usize)..].to_owned()))
+}
+
+pub(crate) fn read_cid(buf: &[u8]) -> Result<(Cid, u64), Error> {
+    // TODO: Add some checks for cid v0
+    let (version, buf) =
+        unsigned_varint::decode::u64(buf).map_err(|e| Error::ParsingError(e.to_string()))?;
+    let (codec, multihash_with_data) =
+        unsigned_varint::decode::u64(buf).map_err(|e| Error::ParsingError(e.to_string()))?;
+    // multihash part
+    let (_hashcode, buf) = unsigned_varint::decode::u64(multihash_with_data)
+        .map_err(|e| Error::ParsingError(e.to_string()))?;
+    let hashcode_len_diff = multihash_with_data.len() - buf.len();
+    let (len, _) =
+        unsigned_varint::decode::u64(buf).map_err(|e| Error::ParsingError(e.to_string()))?;
+
+    let cid: Cid = Cid::new(
+        cid::Codec::from(codec)?,
+        cid::Version::from(version)?,
+        cid::multihash::Multihash::from_bytes(
+            multihash_with_data[0..=(len as usize + hashcode_len_diff)].to_vec(),
+        )?,
+    );
+    let len = cid.to_bytes().len() as u64;
+    Ok((cid, len))
```

Comment:
It doesn't make sense to me why you convert the reader into a slice, then to use it here as a reader. This seems like logic that should be attached to the Cid crate. Thoughts on making this take in a reader and attaching it to Cid crate? You wouldn't need that ld_read function or all this indexing that could panic the program on a bad input

Reply:
Added a todo for read_Cid using a bufreader